### ISSUE TITLE
Stored Solr Fields as Recommendation

### ISSUE THREAD

**Issue by terrellt** [projecthydra/active_fedora#778]
Currently the recommendation, schema.xml, and documentation points towards indexing content as non-stored fields. I assume the reason is to push for the smallest possible index. However, Hydra applications are now needing to do atomic updates of solr documents, which has a requirement that all fields are stored (https://wiki.apache.org/solr/Atomic_Updates#Stored_Values). Spotlight has this requirement, and the work the Applied Linked Data group is doing is leading towards atomic updating of linked data in the solr index.

Can we change the schema.xml to only have stored fields, and change documentation to support it?

**Comment by awead** [comment]
@terrellt I don't think we'd need to change the schema.xml, but just the solr fields themselves, such as using `:stored_searchable` for everything. I believe it's just facet fields, indexed as `:facetable`, that are indexed but not stored.

**Comment by jcoyne** [comment]
Adam

We don't want to change the Solrizer definitions, because that could cause
current application to break.  What we decided was to gradually encourage
people to not use non-stored fields.  So by removing them from the
schema.xml it will only affect people building new applications.  We also
need to update documentation to encourage people to avoid :searchable and
:facetable and to use :stored_searchable or :symbol.  You should feel free
to use non-stored fields if you need to, but understand that you are
cutting yourself off from some of the benefits of Solr's atomic updates.

-Justin

**Comment by awead** [comment]
I'm full support of stored-only fields. It makes things a lot easier. Might you just amend the suffixes of existing Solrizer definitions to use stored fields? For example, `_sim` "string, indexed, multivalued" becomes `_ssim` "stored, string, indexed, multivalued." The current schema.xml file is generalized enough to allow this.

**Comment by awead** [comment]
Alternatively, you could change the schema.xml file so that `stored=true` is for most terms, thereby making differences between `_sim` and `_ssim` irrelevant.

**Comment by jcoyne** [comment]
@awead I think your suggestion of changing the solrizer definitions would break applications where they are using `Solrizer.solr_name(:searchable)` to get back a key for looking up a value in a solr document, which is now the wrong key.  I think changing the schema to stored=true would be odd because there would be two suffixes that mean exactly the same thing. Why not just let it fail fast when they use an undefined key? Then if they are really sure they want that, then they can add it back into place.

**Comment by awead** [comment]
so you're suggesting using completely new suffixes?

**Comment by jcoyne** [comment]
I'm saying we just remove suffixes that aren't stored (e.g. `_teim`, `_sim`, `_dtim`, `_tei`, `_si`, `_dti`, etc) from the schema.xml

**Comment by scande3** [comment]
Quick question: for fields that Active Fedora uses by default that aren't stored (I believe active_fedora_model_ssi, object_state_ssi, system_modified_dtsi, and system_create_dtsi are all of them), would one define those fields specifically in the schema.xml then? Mostly just a note that either not all of the suffixes can be removed unless those default Active Fedora fields are added to schema.xml.

(Disclaimer: That list is from Active Fedora 7/8 and I'm unsure if those fields still exist in that form in Active Fedora 9).

**Comment by terrellt** [comment]
@scande3 Good point about those! Yeah, I'd define them explicitly.

**Comment by jcoyne** [comment]
@scande3 these are all stored: active_fedora_model_ssi, object_state_ssi, system_modified_dtsi, and system_create_dtsi  <-Note the `si` at the end?

**Comment by scande3** [comment]
@jcoyne - oops, yes, you are right. I have no idea why I thought they weren't.

**Comment by atz** [comment]
Why would you ever care to remove any of these dynamic namespaces that are logically just part of a matrix of field options on/off?  Them being there unused costs basically nothing.  You cannot reasonably  exclude downstream use cases for unstored fields unless you believe that all hydra instances *must* use atomic updates.  Or they *must* support atomic updates even if they don't use them?  That seems overstated to me.  

Side note: `:symbol` is the worst flag in Solrizer's catalog of contrived terminology.  In the apps I observe and maintain, the interjection of Solrizer is an almost uniform source of confusion and bloat.  `:facetable` also does not do what the several generations of developers who came before me intended, in particular when I see it applied to date and string fields.  But @jcoyne is right, we absolutely cannot change Solrizer's default behavior **again**.  

It would really help to have OM allow you to specify the exact field name(s) being targeted during `set_terminology` rather than go through solrizer's indirection.  Then you could know *in your code* whether you were using any fields that ran afoul of the new modeling.

**Comment by dchandekstark** [comment]
-1. This issue seems to me a symptom of fundamental design questions related to indexing in ActiveFedora and Hydra (not to mention Fedora itself) and needs more context and conversation to understand the full implications. How will this affect full text discovery? Will AF/Hydra support updating multiple indexes out of the box?

**Comment by terrellt** [comment]
@atz Largely because it's the only clean way to make a recommendation that doesn't break backwards compatibility - it would take work on the dev's part (uncommenting in hydra-jetty's schema.xml) to use unstored fields if they weren't in the schema.

Yes, I hate `:symbol` too. Bad name.

**Comment by terrellt** [comment]
@dchandekstark It doesn't affect full text discovery or do anything to harm currently supported features, no? I don't believe there's a case where a solr feature is unsupported if the field is stored.

**Comment by dchandekstark** [comment]
@terrellt What I mean about full-text discovery is: What are the implications for Solr (and Blakclight) performance of storing, say, millions of pages of text in your Solr index?

**Comment by jcoyne** [comment]
@dchandekstark I don't think we're saying you can't do that if you want to, but Solr doesn't support atomic indexing if you're doing that.  We need atomic indexing for some new features.  You as the implementer need to choose use atomic indexing or use non-stored fields.   We are giving a considered recommendation about which direction we'd like to take this in.

**Comment by dchandekstark** [comment]
@jcoyne I get that.  I just want to be informed about the tradeoffs.  What are the new features?  And am I "on my own" if I choose to use non-stored fields?

**Comment by escowles** [comment]
I agree with @dchandekstark -- I think lots of people store fulltext in non-stored fields, and would like to know what the performance impact of converting those to stored fields would be.

**Comment by atz** [comment]
Vufind, for example, writes fulltext to two fields, but neither one is stored:
https://github.com/vufind-org/vufind/blob/master/solr/biblio/conf/schema.xml#L117-L118

Blacklight doesn't have an explicit field for fulltext, [but warns](https://github.com/projectblacklight/blacklight-jetty/blob/master/solr/blacklight-core/conf/schema.xml#L33-L35):
```
set stored="false" for all fields possible (esp large fields) when you only need to search on the field but don't need to return the original value.
```

In general forcing Solr to store large documents will degrade performance very quickly.  Solr is not a document/file store and using it like one is anti-pattern.  

In practice, we do not need the whole document text **en bloc** at the time we are asking the *index* what documents are relevant.  At query time we only need the *index* of the fulltext.  Similarly, there is no need for direct retrieval of the full document competing with actual searches for Solr resources (instead route those to a proper datastore).  Because if you consider that need legit, a similar line of thought would say, OK, why not store the PDF or TIFF of the whole document as binary encoded field in the solr document?  That would be wildly inappropriate!  

If I may turn the question around, who do you imagine as the consumer for unpaged single-shot fulltext in the response at query time?  And if the use isn't part of searching, it is easily replaced with a proper datastore.

tl;dr: the size of the new index/datastore would be orders of magnitude larger than the previous version.

**Comment by jcoyne** [comment]
@atz I don't see anyone as the consumer of the stored field except for Solr itself. The reason I want to go this way is for performance reasons.  Solr doesn't allow you to do atomic updates unless all fields in the document are stored.  Software such as Spotlight (for Blacklight) requires atomic indexing and other applications such as linked data could benefit from atomic updates too.

**Comment by jcoyne** [comment]
My back of the envelope calculation based on _The Hound Of the Baskervilles_

```
128 pages (according to Amazon)
338kb in text (according to Project Gutenberg)
2.64kb averge of per page
2.6gb per million pages of OCR text.
```
That's in UTF-8. I think its fair to assume that Solr uses a more compact format for storing the data in the index.

**Comment by dchandekstark** [comment]
@jcoyne It seems that you are only talking about *indexing* performance, and not considering *searching* performance (assuming a single-index architecture).  Stored fields is one of two major performance areas related to schema design given at https://wiki.apache.org/solr/SolrPerformanceFactors.  The Lucene wiki also mentions limiting stored fields as a means of improving searching speed (http://wiki.apache.org/lucene-java/ImproveSearchingSpeed).  

Certainly we should do what we can to make the write operations of the system performant. I would just like more information that this recommendation will not degrade searching performance.

**Comment by jcoyne** [comment]
@dchandekstark both of those pages indicate that "retrieving" those fields is expensive. I have no intention of retrieving a full text field. I'm pretty sure setting `fl` is enough to mitigate that.

**Comment by dchandekstark** [comment]
@jcoyne By my reading of https://wiki.apache.org/solr/SolrPerformanceFactors, storing large fields can affect performance even if you are not retrieving them, because "the higher byte count, the sparser the documents will be distributed on disk and more I/O is necessary to retrieve the fields (usually this is a concern when storing large fields, like the entire contents of a document)".

**Comment by escowles** [comment]
I just checked my Solr index and found that we are already indexing our fulltext in a stored field (fulltext_tesim) and that about 15% of our records have fulltext populated (the vast majority of our PDFs, which are mostly newspapers and other short documents).  There are a small number of non-stored fields defined in our schema (*spell and *_sim) , but I'm much less worried about those fields increasing the index/record size.

So I think I'm happy that this isn't going to have a big impact on performance for us.  Sites with a larger number of text documents (or larger documents) may have issues though.

**Comment by dchandekstark** [comment]
@jcoyne @awead @escowles @terrellt All I'm saying is that it's not clear in the general case that this is a benefit to the Hydra stack.  If we can work out the details so that this does not create headaches for folks who index large amounts of text, I will feel better about moving forward with the recommendation.

**Comment by mjgiarlo** [comment]
Do we know anyone who indexes large amounts of text who might be willing to run some benchmarks for us?

Because I want to see Spotlight integrate into more Hydra apps, I'd like to see this work pursued but I think @dchandekstark has raised some worthy concerns that bear some investigation.

**Comment by escowles** [comment]
I was planning on bringing this up at this week's Hydra Tech call -- I can ask if anyone has a lot of text resources who can do some testing.

I would also like to see more info on the need for atomic updates by Spotlight or other apps.  If a site wanted to have non-stored fields, couldn't they just use an async queue system like Sufia?  Would Spotlight even need to update the records for the objects with indexed fulltext?  Some of us want Spotlight to store all of its metadata in Fedora objects anyway, could that be part of an alternate workflow that would avoid the need for all-stored fields?

**Comment by terrellt** [comment]
@escowles You could store and reindex when necessary, but the applied linked data group has use cases which would only be efficient if the index could update independently of interacting with the repository.

**Comment by cbeer** [comment]
Out-of-the-box, Spotlight ships with a configuration that uses atomic updates (and a reasonable person could come along and implement other types of indexing strategies). Atomic updates are a nice default because:

- they get Spotlight out of the business of caring about your index, how you build it, etc. Spotlight can apply its enhancements on top of whatever you can provide.

- they are quick. If we tried to reindex an object from scratch every time Spotlight had something to add to the index, it would take multiple seconds, at least. With Atomic Updates, we get the data into the index quickly and the curator sees results instantly (and, sure, there are hybrid solr + db strategies that might work similarly, but they tend to introduce significant complexity especially around updating information)

It seems like linked data use cases may share similar concerns (although an asynchronous system may be an equally good fit), and being able to treat linked data concerns as an enhancement on top of your index may also be more flexible, reusable, and easier to get to grips with.

**Comment by escowles** [comment]
@cbeer, I see the appeal of working directly with Solr, for both the speed and simplicity reasons you outline.  Based on my own repository using a stored field for fulltext indexing, I strongly suspect the performance concerns are overblown.

But I would like to understand the Spotlight persistence better.  I've been assuming that Spotlight (and linked data caching) would write their own Solr records, so reindexing records when there were updates in Fedora (or whatever) wouldn't clobber them.  If so, then it seems like they could be in a separate Solr index from the object records with fulltext indexing, allowing them to peacefully co-exist.

But is that how it really works?  Is Spotlight actually updating the Solr records for the objects instead?  Is some other approach used to avoid overwriting those updates (more atomic updates, maybe)?

**Comment by dchandekstark** [comment]
I feel satisfied that my concerns have been acknowledged, so consider me a **+0** now.  The issues I raised are general in nature and not based on personal observation or specific usage at my institution.  Mainly I just don't want us to paint ourselves into a corner on this, or create problems for folks who have legitimate reasons to use non-stored fields.  There should be a supported path for those cases -- which it seems there will be if this a recommendation and not a requirement.

**Comment by cbeer** [comment]
@escowles Spotlight is updating Solr records for object, possibly objects indexed externally. If you have an indexing process that wants to come along and clobber the updates, then you probably want to use some of our fine tooling around that (which, if you want to open a sul-dlss/spotlight ticket, we can discuss there.).

**Comment by escowles** [comment]
Based on https://github.com/sul-dlss/spotlight/issues/1152#issuecomment-101344683, I'm confident it wouldn't be too onerous to implement a non-atomic indexing approach if somebody couldn't use all stored fields.

**Comment by jcoyne** [comment]
UCSD just ran into big trouble with having everything stored because their full_text documents were a couple hundred megabytes each. Drawing the index page took over 20s.
